### PR TITLE
feat: [cherry-pick] Major compaction

### PR DESCRIPTION
**Pull request by wayblink**
This PR cherry-picks the following commits:
  fix: speed up segment lookup via channel name in datacoord (#33530) needed by the next commit
  feat: Major compaction (#33620)

issue: #30633
pr: #33620

### CODE REVIEW

**Comment by wayblink** [comment]
rerun ut

**Comment by wayblink** [comment]
rerun ut

File: internal/datacoord/compaction.go [hunk 1]
```diff
@@ -39,7 +39,6 @@ import (
 	"github.com/milvus-io/milvus/pkg/util/typeutil"
 )
 
-//go:generate mockery --name=compactionPlanContext --structname=MockCompactionPlanContext --output=./  --filename=mock_compaction_plan_context.go --with-expecter --inpackage
```
Comment (czs007) [inline internal/datacoord/compaction.go:42]: keep this.
you can modify this in next pr.

Comment (wayblink) [reply internal/datacoord/compaction.go]: updated

**Pull request merged**